File: Solutions/0704_BinarySearch.cpp
class Solution {
public:
    int modSearch(vector<int>& nums, int target, int index) {
        if(nums.size() == 0)
            return -1;

        if(nums.size() == 1) {
            if(nums[0] == target)
                return index;
            else
                return -1;
        }

        if(nums.size() == 2) {
            if(nums[0] == target)
                return index;
            else if (nums[1] == target)
                return index + 1;
            else
                return -1;
        }

        int mid = nums.size() / 2;
        printf("nums.size() = %d, mid = %d, index = %d\n", nums.size(), mid, index);

        if(target == nums[mid]) {
            return index + mid;
        } else if(target == nums[mid + 1]) {
            return index + mid + 1;
        } else if (target < nums[mid]) {
            vector<int> LHS(nums.begin(), nums.begin() + mid);
            return modSearch(LHS, target, index);
        } else {
            vector<int> RHS(nums.begin() + mid, nums.end());
            return modSearch(RHS, target, index + mid);
        }
    }

    int search(vector<int>& nums, int target) {
        return modSearch(nums, target, 0);
    }
};

File: Solutions_Easy/0704_BinarySearch.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        ptr_left = 0
        ptr_right = n - 1

        while ptr_right - ptr_left > 1:
            mid = (ptr_right + ptr_left) // 2
            print(f"{ptr_left=},{ptr_right=},{mid=}")
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                ptr_left = mid
            else:
                ptr_right = mid

        if nums[ptr_right] == target:
            return ptr_right
        if nums[ptr_left] == target:
            return ptr_left
        return -1
